### PR TITLE
Add ability to restrict add-on detail/version API and downloads in specific regions

### LINKED ISSUE #15942: Restrict add-on detail/versions API and download views for add-ons in specific regions
addons-server companion to https://github.com/mozilla/addons-frontend/issues/9842, we need some add-ons to be restricted in some specific regions.

addons-frontend and nginx (*) will pass a `X-Country-Code` header to API requests to the add-on detail and versions API requests indicating the region the client request is originating from, in [ISO 3166-1 alpha-2](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2) format. We need to add a table containing add-on region exclusions for a given add-on, probably just stored as flat json list of country codes. When an add-on is requested from a restricted country, we need to return a HTTP 451 response instead of the normal detail page response.

This same logic also needs to be added to download views.

As a follow-up, we'll likely need something to be added to be able to bypass this if logged in as an admin, while adding something to the response to indicate to the frontend that we bypassed the restriction, but this doesn't have to be implemented right away.

(*) We will have to let clients override the header, because of addons-frontend server side rendering. We may have some config in nginx to only allow that under specific circumstances though, to prevent clients from playing with this.

### PR DESCRIPTION
**Pull request by diox**
Fixes #15942

Depends on nginx config change to automatically pass the `X-Country-Code` header (and addons-frontend changes to pass that same header when doing server side rendering requests, as well as handling the 451 errors)

### CODE REVIEW

File: src/olympia/api/middleware.py [hunk 1]
```diff
@@ -2,16 +2,22 @@
...
+class APIRequestMiddleware(MiddlewareMixin):
     def identify_request(self, request):
         request.is_api = re.match(settings.DRF_API_REGEX, request.path_info)
 
     def process_request(self, request):
         self.identify_request(request)
 
+    def process_response(self, request, response):
+        if request.is_api:
+            patch_vary_headers(response, ['X-Country-Code'])
```
Comment (eviljeff) [inline src/olympia/api/middleware.py:18]: _all_ api requests?

File: src/olympia/api/exceptions.py [hunk 2]
```diff
@@ -4,13 +4,20 @@
...
 from rest_framework import exceptions
 from rest_framework import status
 from rest_framework.response import Response
 from rest_framework.views import set_rollback
 
 
+class UnavailableForLegalReasons(exceptions.APIException):
+    status_code = status.HTTP_451_UNAVAILABLE_FOR_LEGAL_REASONS
+    default_detail = _('Unavailable for legal reasons.')
+    default_code = 'unavailable_for_legal_reasons'
```
Comment (eviljeff) [inline src/olympia/api/exceptions.py:18]: add the Link header as suggested?  https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/451  (it's a "SHOULD" in https://tools.ietf.org/html/rfc7725#section-4)

File: src/olympia/addons/views.py [hunk 3]
```diff
@@ -224,6 +230,13 @@ def check_object_permissions(self, request, obj):
                 'is_disabled_by_mozilla': obj.status == amo.STATUS_DISABLED,
             }
             raise exc
+        region_code = (
+            self.request and self.request.META.get(
+                'HTTP_X_COUNTRY_CODE', None))
+        if region_code and AddonRegionalRestrictions.objects.filter(
+                addon=obj,
+                excluded_regions__contains=region_code.upper()).exists():
```
Comment (eviljeff) [inline src/olympia/addons/views.py:238]: Once we start getting this for all requests it'll probably be more efficient to use `select_related` on `'regional_restrictions'` vs. a separate query (as a follow-up if so)

File: src/olympia/addons/tests/test_views.py [hunk 4]
```diff
@@ -251,6 +252,17 @@ def test_get_not_listed(self):
         assert data['is_disabled_by_developer'] is False
         assert data['is_disabled_by_mozilla'] is False
 
+        AddonRegionalRestrictions.objects.filter(
+            addon=self.addon).update(excluded_regions=['AB', 'CD', 'FR'])
+        # Regional restrictions should be processed after other permission
+        # handling, so something that would return a 401/403/404 without
+        # region restrictions would still do that.
+        response = self.client.get(self.url)
```
Comment (bobsilverberg) [inline src/olympia/addons/tests/test_views.py:260]: Just a drive-by comment, but where is one of `['AB', 'CD', 'FR']` set as `X-Country-Code` for the request? I.e., what reason do we have to believe the test will care that these have been added to `excluded_regions`?

File: src/olympia/versions/tests/test_views.py [hunk 5]
```diff
@@ -179,6 +188,14 @@ def test_download_for_unlisted_addon_owner(self):
         self.assert_served_internally(
             self.client.get(url), False, attachment=True)
 
+        # Even allowed to bypass georestrictions.
```
Comment (bobsilverberg) [inline src/olympia/versions/tests/test_views.py:191]: Another drive-by question: where are all of the exclusions tested for in this file defined, in terms of requirements?

File: src/olympia/api/exceptions.py (hunk 2 above)
Comment (willdurand) [reply src/olympia/api/exceptions.py:18]: +1

File: src/olympia/versions/views.py [hunk 6]
```diff
@@ -98,14 +99,24 @@ def is_appropriate_reviewer(addon, channel):
...
-        loc = urlparams(file_.get_file_cdn_url(attachment=attachment),
-                        filehash=file_.hash)
-        response = http.HttpResponseRedirect(loc)
-        response['X-Target-Digest'] = file_.hash
+        # When serving the file for the general public through the CDN, we need
+        # to obey regional restrictions
+        region_code = request.META.get('HTTP_X_COUNTRY_CODE', None)
+        if region_code and AddonRegionalRestrictions.objects.filter(
+                addon=addon,
+                excluded_regions__contains=region_code.upper()).exists():
```
Comment (willdurand) [inline src/olympia/versions/views.py:107]: Shouldn't we use `icontains` here as well?

File: src/olympia/addons/views.py (hunk 3 above)
Comment (willdurand) [reply src/olympia/addons/views.py:238]: Also, shouldn't we use `icontains` here?

File: src/olympia/versions/views.py [hunk 7]
```diff
@@ -98,14 +99,24 @@ def is_appropriate_reviewer(addon, channel):
...
-        # Content-Disposition: attachment for attachments. To work around this,
-        # if attachment=True, get_file_cdn_url() changes the path to something
-        # we recognize in the nginx config.
-        loc = urlparams(file_.get_file_cdn_url(attachment=attachment),
-                        filehash=file_.hash)
-        response = http.HttpResponseRedirect(loc)
-        response['X-Target-Digest'] = file_.hash
+        # When serving the file for the general public through the CDN, we need
+        # to obey regional restrictions
+        region_code = request.META.get('HTTP_X_COUNTRY_CODE', None)
```
Comment (willdurand) [inline src/olympia/versions/views.py:104]: It's interesting that we call the header "country code" but the variable is "region code".

**Review by willdurand — APPROVED** [review]
r+wc

File: src/olympia/api/middleware.py (hunk 1 above)
Comment (diox) [reply src/olympia/api/middleware.py:18]: It's just a `Vary` on a header that shouldn't be sent very often, so I think it's safe

Comment (diox) [reply src/olympia/api/middleware.py:18]: It's a lot easier to add there - we might return the 451 for anything that depend on an add-on, but we don't have a centralized way of returning the `Vary` header otherwise

File: src/olympia/addons/views.py (hunk 3 above)
Comment (diox) [reply src/olympia/addons/views.py:238]: Doing `icontains` would make the query a ` LIKE %` (which could lead to false positives) instead of a `JSON_CONTAINS()`

File: src/olympia/addons/tests/test_views.py (hunk 4 above)
Comment (diox) [reply src/olympia/addons/tests/test_views.py:260]: It's a good point, I manually passed the header for `test_addon_regional_restrictions` below but forgot it there (and in a couple other places). I'll fix that.

File: src/olympia/versions/tests/test_views.py (hunk 5 above)
Comment (diox) [reply src/olympia/versions/tests/test_views.py:191]: What do you mean ?

File: src/olympia/versions/views.py (hunk 6 above)
Comment (diox) [reply src/olympia/versions/views.py:107]: https://github.com/mozilla/addons-server/pull/15965/files#r520608452

File: src/olympia/versions/tests/test_views.py (hunk 5 above)
Comment (bobsilverberg) [reply src/olympia/versions/tests/test_views.py:191]: I mean, who says that we should be doing this exclusion? Who says that it's okay for an owner in a restricted region to be able to download their add-on? (Or maybe I'm just not understanding what's being tested here)

File: src/olympia/versions/views.py (hunk 7 above)
Comment (diox) [reply src/olympia/versions/views.py:104]: Yeah. We called the header country code because it follows https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2 but I didn't want to call the model/field country because it was a bit of an oversimplification IMHO - technically, a different country code would be sent for special collectivities and territories, even if it's the same "country".

For instance, Guadeloupe has its own ISO 3166-1 alpha-2 country code, despite being part of France.

File: src/olympia/addons/views.py (hunk 3 above)
Comment (willdurand) [reply src/olympia/addons/views.py:238]: Interesting. how do we ensure that entries will be correct in the JSON then?

Comment (eviljeff) [reply src/olympia/addons/views.py:238]: a big stick gestured at whoever updates the data 😜

File: src/olympia/api/exceptions.py (hunk 2 above)
Comment (diox) [reply src/olympia/api/exceptions.py:18]: I can look into that, but I'm not sure where to link to. https://www.mozilla.org/en-US/about/policy/transparency/ ?

Comment (eviljeff) [reply src/olympia/api/exceptions.py:18]: Sounds good.

File: src/olympia/versions/tests/test_views.py (hunk 5 above)
Comment (diox) [reply src/olympia/versions/tests/test_views.py:191]: It's not defined anywhere so I made the call myself, mostly to make it easier to implement a bypass for admins & reviewers. I don't think it matters much, and it matches what we allow for disabled, so that makes sense.

File: src/olympia/addons/views.py (hunk 3 above)
Comment (diox) [reply src/olympia/addons/views.py:238]: Something to ensure in https://github.com/mozilla/addons-server/issues/15968 in the long run

File: src/olympia/api/exceptions.py (hunk 2 above)
Comment (diox) [reply src/olympia/api/exceptions.py:18]: ffb10974610e186d7f8cfeff99de2f88d4bfd01b

File: src/olympia/addons/tests/test_views.py (hunk 4 above)
Comment (diox) [reply src/olympia/addons/tests/test_views.py]: 7226d9cd3e49743f4c28b8286a7f8e4ca6aff569
